fontconfig: 2.14.0 -> 2.14.2

Scrumplex opened the pull request:
###### Description of changes

Supersedes #197523

Changelog: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/compare/2.14.0...2.14.2

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

r-ryantm committed:
fontconfig: 2.14.0 → 2.14.2

https://gitlab.freedesktop.org/fontconfig/fontconfig/-/compare/2.14.0...2.14.2

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
nixos/fontconfig: refactor antialias option for fontconfig 2.14.1

`sub-pixel` has been enabled by default since 2.14.1: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/2b6afa02ab2b7dd3796a48cf47896c4c6de4d6ba
`antialias` since 2.14.1: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/0825a178e82c58ae97e908d49e5ad7abe88ed7f5
`lcdfilter` since 2.13.95: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/e1c7c6d7442ace9c20c424265880ff57c23881f8
`hintstyle` since 2.12.1: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/98434b3392172233094cac25ade7225c93da9f1c

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

jtojnar reviewed (changes requested):
Thanks, note that the same thing applies for `cfg.subpixel.rgba` and `10-sub-pixel-$(PREFERRED_SUB_PIXEL_RENDERING).conf`:

- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-sub-pixel-bgr.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-sub-pixel-none.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-sub-pixel-rgb.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-sub-pixel-vbgr.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-sub-pixel-vrgb.conf

jtojnar commented:
Also this needs to target staging branch. Please click _Convert to draft_ before changing the branch to prevent GitHub from pinging everyone if too new merge base is used.

Scrumplex commented:
> Thanks, note that the same thing applies for `cfg.subpixel.rgba` and `10-sub-pixel-$(PREFERRED_SUB_PIXEL_RENDERING).conf`:

Argh, I was scrolling through all the changed files and didn't see that `10-sub-pixel-$(PREFERRED_SUB_PIXEL_RENDERING).conf` is being installed as a default config now

jtojnar reviewed (commented):
Changes look good from what I skimmed from changelog, did not test.

jtojnar commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -14,13 +14,13 @@
   src = fetchurl {
     url = "https://www.freedesktop.org/software/fontconfig/release/${pname}-${version}.tar.xz";
-    sha256 = "3L64TJx0u/2xM9U1/hx77cnyIhqNrzkUuYTETFIOm6w=";
+    sha256 = "sha256-26aVtXvOFQI9LO7e+CBiwrkl5R9dTMSu9zbPE/YKRos=";
```
Slightly redundant:

```suggestion
    hash = "sha256-26aVtXvOFQI9LO7e+CBiwrkl5R9dTMSu9zbPE/YKRos=";
```

jtojnar commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -67,12 +67,6 @@ let
-          <const>${cfg.subpixel.rgba}</const>
-        </edit>
         <edit mode="append" name="lcdfilter">
           <const>lcd${cfg.subpixel.lcdfilter}</const>
```
Ugh, looks like `lcdfilter` has been linked [since 2.13.95](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blame/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/meson.build#L51):

- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/11-lcdfilter-default.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/11-lcdfilter-legacy.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/11-lcdfilter-light.conf

And `hintstyle` [since 2.12.1](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commits/2.12.1):

- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-hinting-full.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-hinting-medium.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-hinting-none.conf
- https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/7861a719616b4b132b9cac089c6c64f47832edb1/conf.d/10-hinting-slight.conf

Scrumplex replied:
Hmm. The latter will require a breaking change, as the files are named `10-hinting-{full,medium,none,slight}.conf` while the enum for `hinting.style` has the following values `[ "hintnone" "hintslight" "hintmedium" "hintfull" ]`

Alternatively, we could create a substring of these values to avoid a breaking change, which would be less clean in my opinion.

Scrumplex replied:
Yep. That apply function seems like it's a nicer solution.

Scrumplex commented:
I haven't tested my last changes yet

jtojnar commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -183,6 +178,26 @@ let
+      (replaceDefaultConfig "11-lcdfilter-default.conf" "11-lcdfilter-${cfg.subpixel.rgba}.conf")
+    }
+
+    ${optionalString (cfg.subpixel.lcdfilter == "none") ''
```
Hmm, it’s weird that there is no `11-lcdfilter-none.conf`. I wonder if it is fine to not define it.

Scrumplex replied:
I just pushed my commits again. This patch is now backported and the NixOS module doesn't have this special check anymore.

jtojnar commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -350,17 +365,25 @@ in
+                val' = lib.removePrefix "hint" val;
+                warning = "The option `${from}` contains a deprecated value `${val}`. Use `${val'}` instead.";
+              in
+              lib.warnIf (lib.hasPrefix "hint") warning val';
```
Sorry, should be `lib.hasPrefix "hint" val`

Scrumplex replied:
done

jtojnar commented on nixos/modules/config/fonts/fontconfig.nix:
Oh and perhaps one more commit:

```
nixos/fontconfig: Change default subpixel to none

This is the upstream default since 2.14.2: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/030759b74f5b3ce7fab6d17bbda6377444e82841
```

Scrumplex replied:
done

Scrumplex committed:
nixos/fontconfig: Change default antialiasing style to greyscale instead of subpixel

fontconfig before version 2.13.1 was apparently implicitly not using
subpixel antialiasing. The fontconfig NixOS module deviated from this,
using subpixel antialiasing with `rgb` layout by default. In fontconfig
2.14.1, subpixel antialiasing was inadvertently enabled as the default:
https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/2b6afa02ab2b7dd3796a48cf47896c4c6de4d6ba

According to https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/337,
that deviates from GNOME/GTK’s defaults, which resulted in apps taking the
settings directly from fontconfig (e.g. Firefox) from diverging from GNOME
programs.

The change was subsequently reverted in 2.14.2, choosing the greyscale
antialiasing explicitly: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/030759b74f5b3ce7fab6d17bbda6377444e82841
Let’s reflect this default setting in the NixOS module.

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>
Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
Should I add some entries to the 23.05 release notes for some of these changes?

jtojnar reviewed (approved):
Did not test but changes look okay.

vcunat commented:
FYI, if this should make it to 23.05, it would better get reviewed and merged quite soon.  (ideally by yesterday)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/52

RaitoBezarius commented:
I think we should merge this for post-23.05. fontconfig stuff has already been quite a ride.

jtojnar commented:
Now that branch off has happened, I have rebased this and moved the announcement to 23.11 release notes.

trofi commented:
I suspect b5d2d701d1e983fe82770905e8fce703ba38caa3 ` nixos/fontconfig: refactor antialias option for fontconfig 2.14.1` broke `system` generation for me:

```
```

trofi commented:
Looks like having `fonts.fontconfig.subpixel.rgba = "rgb";` alone is enough to trigger config generation failure.

trofi commented:
Proposed `dst` fix as https://github.com/NixOS/nixpkgs/pull/235901

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/evolution-font-changes-with-23-11/37398/3
